client/commonplugins/commonplugins.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package commonplugins
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"os/exec"
+	"syscall"
+	"time"
+
+	"github.com/hashicorp/go-version"
+)
+
+const (
+	CPI_TIMEOUT_SOFT time.Duration = 2 * time.Second
+	CPI_TIMEOUT_HARD time.Duration = 1 * time.Second
+)
+
+var (
+	ErrPluginNotExists     error = errors.New("plugin not found")
+	ErrPluginNotExecutable error = errors.New("plugin not executable")
+)
+
+type CommonPlugin interface {
+	Fingerprint(ctx context.Context) (*PluginFingerprint, error)
+}
+
+// CommonPlugins are expected to respond to 'fingerprint' calls with json that
+// unmarshals to this struct.
+type PluginFingerprint struct {
+	Version *version.Version `json:"version"`
+	Type    *string          `json:"type"`
+}
+
+// runPlugin is a helper for executing the provided Cmd and capturing stdout/stderr.
+// This helper implements both the soft and hard timeouts defined by the common
+// plugins interface.
+func runPlugin(cmd *exec.Cmd) (stdout, stderr []byte, err error) {
```

Comment:
The  `exec.Cmd` we're passing in here already has a context, which makes ownership of the command timeout ambiguous. If the caller closes the context for the command it'll SIGTERM the plugin process. Whereas the command timeout here is unrelated to that context. Maybe we want to require that the caller doesn't set a timeout in the context but only passes the timeout and the context for agent shutdown? And then we'd could make the command timeout a child context of that.

Having the caller own the timeout would allow each command to have different timeouts. Ex. we have a 5s timeout for fingerprinting host volume plugins but a 60s timeout for creating host volumes.